subprojects/build-events/src/main/java/org/gradle/internal/build/event/types/DefaultFailure.java

```diff
@@ -47,15 +62,44 @@ public String getDescription() {
 
     @Override
     public List<? extends InternalFailure> getCauses() {
-        return cause == null ? Collections.emptyList() : Collections.singletonList(cause);
+        return causes;
+    }
+
+    @Override
+    public List<InternalBasicProblemDetailsVersion3> getProblems() {
+        return problems;
     }
 
     public static InternalFailure fromThrowable(Throwable t) {
+        return fromThrowable(t, Collections.emptyMap(), p -> null);
+    }
+
+    public static InternalFailure fromThrowable(Throwable t, Map<Throwable, Collection<Problem>> problemsMapping, Function<Problem, InternalBasicProblemDetailsVersion3> mapper) {
         StringWriter out = new StringWriter();
         PrintWriter wrt = new PrintWriter(out);
         t.printStackTrace(wrt);
         Throwable cause = t.getCause();
-        InternalFailure causeFailure = cause != null && cause != t ? fromThrowable(cause) : null;
-        return new DefaultFailure(t.getMessage(), out.toString(), causeFailure);
+        List<InternalFailure> causeFailures;
```

Comment:
🤔 This might worth a comment here, detailing quickly what this mapping does, so if somebody digs around here can understand faster if this code is important for them or not.